### PR TITLE
Flight State Updates and API Integration

### PR DESCRIPTION
**Pull request by rdoinhs**
### Description
This PR integrates the ability to search and confirm flights through the Duffel API, restructures and defines the flow of Flight components and includes miscellaneous updates to the Card Component, animations, and Input component memoization. State is defined locally in a separate component context and has not yet been integrated into the Itinerary flow. 

⚠  **This change requires an update to the .env file** ⚠

#### Supported Features ✔️
* Searching for one-way and round-trip flights, both international and domestic.
* Limiting offers based on the availability of cabin class.
* Sorting offers by airline (for departing offers), lowest price, and shortest duration (as a client-side transformation).
* Viewing the total price for round-trip offers, duration, number of stops, and multiple origin and destination airports based on the IATA location code or city name.
* Editing initial search parameters prior to confirmation.
* Viewing/Modification of departing and returning flight states without loss of flight selection, should the choice of offer remain the same.
* Loading offer items using pseudo pagination in the form of an infinitely scrolling list.

#### Incomplete Features / List of Known Issues / Code Refactoring Opportunities ⚠️
- Errors in server components do not surface relevant client-side alerts or notifications. This could occur due to several edge cases including, but not limited to:
  * The origin or destination is not a valid IATA code or city name, where a sufficiently incorrect name cannot be fixed by a request to the suggestions endpoint.
  * The time zone to or from the local time is ahead or behind by a number of days. For example, currently, attempting to search for a flight to New York at 10:00 PM Friday Pacific Standard Time will fail to advance to a list of results (as it would now be Saturday morning in NYC). This is particularly relevant for flights to international locations.
* Sorting by result type may be inaccurate due to unloaded items not being factored into client-side sorting at the time of effect call. This is a flaw in the design of state management.
* Multi-stop routes do not show intermediate airports (they are factored into total duration, however).
* The number of passengers has no effect on returned results.
* The infinitely scrolling list may remain fixed in a loading state due to insufficient notification of parent component scroll height adjustment. Workaround: scroll up and down again. Parameters need adjustment.
* State management for the Presentations/Select.tsx component and the overall state management solution is overly complicated and needs modularization of shared interests, potentially breaking things into several sub-contexts and subcomponents.
* Card Component height jumps rapidly between transitions of differing content heights.
* State is inconsistently saved/retrieved on reload of the current component through expanding or contracting in card minimization.
* The change departing flight edit component is incomplete and does not update to surface relevant information after choosing a departing flight.
* The change departing flight edit component overflows its parent card.
* Pending state for buttons activates regardless of the intended flow direction (the trigger is dependent on a contentPending field).
* The Added to Itinerary Page is incomplete and does not surface all relevant information nor offer a way to add another flight or view existing reservations.
* Unnecessary component re-renders are commonplace.

### Motivation and Context
This represents a long-standing WIP feature to add the ability to add flights to the event itinerary. The current pr handles state locally and does not yet integrate with the overall application flow outside of its context.

### Screenshots:
Screenshot 2023-11-24 at 10 58 30 PM

_No Offers Available, The Title and Subtitle will always be filled in despite what appears here._

https://github.com/heyitsmass/BayView/assets/38991991/6093ee08-da2b-42b0-8d0c-7ef7b2330b4a

_Video of features presented above / Demonstration of Flow (skip around to get the idea)_

### Additional Notes
A follow-up PR will include updates to the integration of state to allow for the addition of additional flights and resumption of current flow. The Confirmation component is not yet final. Due to time, several features have been limited or scrapped from the original design including the ability to add amenities and seat selection.

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: src/components/HomePage/Card/FlightCard/Presentations/Book.tsx
Comment (heyitsmass) [inline src/components/HomePage/Card/FlightCard/Presentations/Book.tsx:1]: Is this state being called? I see it defined in the steps but there's no explicit SET_STEP - only two for step 5 in Presentations presumably where step 4 would be?

File: src/components/HomePage/TopBar/index.tsx
Comment (heyitsmass) [inline src/components/HomePage/TopBar/index.tsx:1]: Session is required for supertokens

File: src/components/HomePage/Card/FlightCard/Presentations/Book.tsx
Comment (rdoinhs) [reply src/components/HomePage/Card/FlightCard/Presentations/Book.tsx:1]: At the moment, no. It was/is intended to be the state for managing edits to existing flight reservations.

**Review by heyitsmass — APPROVED** [review]

**Review by cpilande — APPROVED** [review]
Looks good

**Pull request merged**